// @asamuzakjp/css-color/dist/esm/js/cache.js
//#region src/js/cache.ts
var MAX_CACHE = 4096;
/**
* CacheItem
*/
var CacheItem = class {
	#isNull;
	#item;
	/**
	* constructor
	*/
	constructor(item, isNull = false) {
		this.#item = item;
		this.#isNull = !!isNull;
	}
	get item() {
		return this.#item;
	}
	get isNull() {
		return this.#isNull;
	}
};
/**
* NullObject
*/
var NullObject = class extends CacheItem {
	/**
	* constructor
	*/
	constructor() {
		super(Symbol("null"), true);
	}
};
/**
* Generational Cache implementation
*/
var GenerationalCache = class {
	max;
	current;
	old;
	constructor(max) {
		this.max = Math.ceil(max / 2);
		this.current = /* @__PURE__ */ new Map();
		this.old = /* @__PURE__ */ new Map();
	}
	get(key) {
		let value = this.current.get(key);
		if (value !== void 0) return value;
		value = this.old.get(key);
		if (value !== void 0) {
			this.set(key, value);
			return value;
		}
	}
	set(key, value) {
		this.current.set(key, value);
		if (this.current.size >= this.max) {
			this.old = this.current;
			this.current = /* @__PURE__ */ new Map();
		}
	}
	has(key) {
		return this.current.has(key) || this.old.has(key);
	}
	delete(key) {
		this.current.delete(key);
		this.old.delete(key);
	}
	clear() {
		this.current.clear();
		this.old.clear();
	}
};
var genCache = new GenerationalCache(MAX_CACHE);
/**
* set cache
* @param key - cache key
* @param value - value to cache
* @returns void
*/
var setCache = (key, value) => {
	if (key) if (value === null) genCache.set(key, new NullObject());
	else if (value instanceof CacheItem) genCache.set(key, value);
	else genCache.set(key, new CacheItem(value));
};
/**
* get cache
* @param key - cache key
* @returns cached item or false otherwise
*/
var getCache = (key) => {
	if (!key) return false;
	const item = genCache.get(key);
	if (item !== void 0) {
		if (item instanceof CacheItem) return item;
		genCache.delete(key);
		return false;
	}
	return false;
};
/**
* helper function to sort object keys alphabetically
* @param obj - Object
* @returns stringified JSON
*/
var stringifySorted = (obj) => {
	const keys = Object.keys(obj).sort();
	if (keys.length === 0) return "";
	return JSON.stringify(keys.reduce((acc, key) => {
		acc[key] = obj[key];
		return acc;
	}, {}));
};
/**
* create cache key
* @param keyData - key data
* @param [opt] - options
* @returns cache key
*/
var createCacheKey = (keyData, opt = {}) => {
	const { customProperty = {}, dimension = {} } = opt;
	if (!keyData || Object.keys(keyData).length === 0 || typeof customProperty.callback === "function" || typeof dimension.callback === "function") return "";
	return `${`${keyData.namespace || ""}:${keyData.name || ""}:${keyData.value || ""}`}::${[
		opt.format || "",
		opt.colorSpace || "",
		opt.colorScheme || "",
		opt.currentColor || "",
		opt.d50 ? "1" : "0",
		opt.nullable ? "1" : "0",
		opt.preserveComment ? "1" : "0",
		String(opt.delimiter || "")
	].join("|")}::${stringifySorted(customProperty)}::${stringifySorted(dimension)}`;
};
//#endregion
export { CacheItem, NullObject, createCacheKey, getCache, setCache };
